codec/codec.go

```diff
@@ -0,0 +1,46 @@
+package codec
```

Comment:
I actually made a version which put this in the grpc package. Because we have a transport package and the constraint that the transport package should not call anything in the grpc package, putting it in the grpc package leads to less-neat code:
the current code makes transport.stream.codec the only place to record codec to be used . The old approach needs to scatter that in different places on upper layer (grpc package) and could lead to unnecessary complexity e.g., transport needs to pass the receive content-type string to the grpc package to do the processing to extract codec info.